frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -192,6 +192,10 @@ TypeVariableSubstitution* TypeInference::unify(const IR::Node* errorPosition,
 
     TypeConstraints constraints(typeMap->getSubstitutions());
     constraints.addEqualityConstraint(destType, srcType);
+    if (auto itv = destType->to<IR::ITypeVar>())
```

Comment:
That's not correct syntax -- controls can't contain method declarations -- but if they could, that's precisely what this change is trying to address: H needs to be bound by the constructor of c, not the call to x.  I'll add another test case that illustrates this with a normal extern rather than one that uses a factory method returning the extern.

Reply:
Hmm -- I think this case should be correct and illustrate the problem:

    extern ext<H> {
        ext(H v);
        void method<T>(H h, T t);
    }

    control c() {
        ext(16w0) x;
        apply {
            x.method(0, 8w0);
        }
    }

but it fails either with out without this change -- it fails to propagate the bound type for H from the constructor to the method call.  It fails even with an explicit `16w0` for the first arg to method!

Reply:
If this case fails it is a bug; we should add it to our test suite.
H should be unified with `bit<16>` and T with `bit<8>`.

Reply:
Actually this program fails because we do not infer type parameters from constructor arguments. If you modify it as follows: `ext<bit<16>>(16w0) x;` it works. The compiler requires all constructors for generics to specify explicitly type parameter values.

Reply:
We do that because in general constructor arguments may not be enough to infer all type parameters. So it is simpler to just ask users to write the type parameters explicitly. I think this is very consistent. It is not an artifact of the implementation traversal order.